### ISSUE TITLE
HIP 4.1.0 breaks the ability of specializing `std::numeric_limits<T>` for non-standard types `T` 

### ISSUE THREAD

**Issue by leofang** [ROCm-Developer-Tools/HIP#2247]
Hi, I am coming from the CuPy project. I would like to bring to your attention that our CI reports build failures against the latest ROCm 4.1.0 (https://github.com/cupy/cupy/pull/4961#issuecomment-805560552):
```
cupy/cuda/cupy_cub.cu:67:7: error: explicit specialization of 'std::numeric_limits<thrust::complex<float>>' after instantiation
class numeric_limits<thrust::complex<float>> {
      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/rocm-4.1.0/llvm/lib/clang/12.0.0/include/__clang_hip_cmath.h:326:1: note: implicit instantiation first required here
__HIP_OVERLOAD1(double, abs)
^
/opt/rocm-4.1.0/llvm/lib/clang/12.0.0/include/__clang_hip_cmath.h:294:44: note: expanded from macro '__HIP_OVERLOAD1'
  __DEVICE__ typename __hip_enable_if<std::numeric_limits<__T>::is_integer,    \
                                           ^                                   
cupy/cuda/cupy_cub.cu:79:7: error: explicit specialization of 'std::numeric_limits<thrust::complex<double>>' after instantiation
class numeric_limits<thrust::complex<double>> {
      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/rocm-4.1.0/llvm/lib/clang/12.0.0/include/__clang_hip_cmath.h:326:1: note: implicit instantiation first required here
__HIP_OVERLOAD1(double, abs)
^
/opt/rocm-4.1.0/llvm/lib/clang/12.0.0/include/__clang_hip_cmath.h:294:44: note: expanded from macro '__HIP_OVERLOAD1'
  __DEVICE__ typename __hip_enable_if<std::numeric_limits<__T>::is_integer,    \
                                           ^                          
```
which worked fine from ROCm 3.5.0 to 4.0.1. 

This specific file [`cupy_cub.cu`](https://github.com/cupy/cupy/blob/master/cupy/cuda/cupy_cub.cu) provides a wrapper over hipCUB when compiled against HIP, and there we need to specialize the comparison rules for Thrust complex in order to comply with NumPy's rule.
```c++
namespace std {
template <>
class numeric_limits<thrust::complex<float>> {
  public:
    static __host__ __device__ thrust::complex<float> max() noexcept {
        return thrust::complex<float>(std::numeric_limits<float>::max(), std::numeric_limits<float>::max());
    }

    static __host__ __device__ thrust::complex<float> lowest() noexcept {
        return thrust::complex<float>(-std::numeric_limits<float>::max(), -std::numeric_limits<float>::max());
    }
};

template <>
class numeric_limits<thrust::complex<double>> {
  public:
    static __host__ __device__ thrust::complex<double> max() noexcept {
        return thrust::complex<double>(std::numeric_limits<double>::max(), std::numeric_limits<double>::max());
    }

    static __host__ __device__ thrust::complex<double> lowest() noexcept {
        return thrust::complex<double>(-std::numeric_limits<double>::max(), -std::numeric_limits<double>::max());
    }
};
}  // namespace std
```
This file is then compiled using `hipcc ... --include hip_runtime.h ...`.

On ROCm 4.1.0, I think somewhere in the code path in `hip_runtime.h`, an implicit instantiation of `std::numeric_limits<T>` was requested to check `is_integer`. But, it precedes our custom specialization, hence leading to the compilation error.

Note there's a circular dependency: in order to specialize `numeric_limits` for complex numbers, we need to import the Thrust headers first. But, in order to resolve the `__host__`/`__device__` specifiers in Thrust, we need to include `hip_runtime.h` either when invoking `hipcc` as shown above, or before `#include <thrust/complex.h>`. One way or another, the implicit instantiation kicks in before our code.

There is no simple fix AFAIK other than kindly asking the upstream to fix, but perhaps I am missing something obvious. Thanks.

**Comment by leofang** [comment]
Minimal reproducer of the compilation error (it needs some headers from CuPy, but does not require to build CuPy):
```
git clone --recursive https://github.com/cupy/cupy.git
cd cupy
hipcc -I/opt/rocm/include -I/opt/rocm/include/hip -I./cupy/core/include/ --include hip_runtime.h --std=c++11 test_rocm_std.cpp -o test_rocm_std
```
test_rocm_std.cpp:
```c++
#include <cupy/complex.cuh>
#include <hip/hip_runtime.h>
//#include <limits>


namespace std {
template <>
class numeric_limits<thrust::complex<float>> {
  public:
    static __host__ __device__ thrust::complex<float> max() noexcept {
        return thrust::complex<float>(std::numeric_limits<float>::max(), std::numeric_limits<float>::max());
    }

    static __host__ __device__ thrust::complex<float> lowest() noexcept {
        return thrust::complex<float>(-std::numeric_limits<float>::max(), -std::numeric_limits<float>::max());
    }
};
}  // namespace std


int main() {
  thrust::complex<float> a = std::numeric_limits<thrust::complex<float>>::max();
  return 0;
}
```

**Comment by yxsamliu** [comment]
Thanks for reducing the test case. That helps us investigate the issue.

The issue happens because cupy/complex.cuh includes some header files which specializes template ::abs<thrust::complex<double>>, which triggers instantiation of std::numeric_limits<thrust::complex<double>> in HIP headers:

https://github.com/cupy/cupy/blob/master/cupy/core/include/cupy/complex/arithmetic.h#L256

Later on, compiler finds

specialization of std::numeric_limits<thrust::complex<double>>, which causes failure, since specialization of std::numeric_limits<thrust::complex<double>> needs to happen before instantiation of std::numeric_limits<thrust::complex<double>>.

The fix is to put specialization of std::numeric_limits<thrust::complex<double>> at https://github.com/cupy/cupy/blob/master/cupy/core/include/cupy/complex/complex.h#L673 .

**Comment by leofang** [comment]
@yxsamliu I don't think it's the reason. The `abs` functions at line 256 (which you quoted) and line 260 call the internal `detail::complex::abs` defined just a few lines above. There is another `abs` function at line 239, which is called by other functions, but they are all in the same namespace `thrust`, so none of them relies on the ADL for `::abs`, and thus none of them would trigger instantiation of `std::numeric_limits<thrust::complex>`. 

I'd like to emphasize that our codebase has been tested across all recent CUDA (8.0-11.2) and ROCm (3.5.0-4.1.0) versions, and only ROCm 4.1.0 throws compilation errors. Therefore I incline to think this is more likely trigged by some changes recently introduced in ROCm/HIP. Do you know what could be the possible change leading to this? Thanks!

**Comment by leofang** [comment]
Another thing is `std::numeric_limits` cannot exist in CuPy's complex headers, see e.g.
https://github.com/cupy/cupy/blob/10a3d9fda11ac90c283976a1bcd1d0e589fd922e/cupy/core/include/cupy/complex/cproj.h#L33
The reason is nvRTC does not support them (though I don't know about hipRTC). ~~So, unfortunately @yxsamliu the solution you suggested is not really applicable...~~

EDIT: I spoke a little too fast. For the purpose of nvrtc, we can wrap it under a `__CUDA_ARCH__` macro check if it's deemed the only solution (which I doubt).

**Comment by leofang** [comment]
@yxsamliu Another note: rocThrust, like Thrust (which CuPy bundles), does not specialize any `std::numeric_limits` in the complex headers:
```
$ git clone https://github.com/ROCmSoftwarePlatform/rocThrust.git
$ cd rocThrust/
$ grep -R ./thrust/detail/complex/ -e numeric_limits
./thrust/detail/complex/cproj.h:    // std::numeric_limits<T>::infinity() doesn't run on the GPU
./thrust/detail/complex/cproj.h:    // std::numeric_limits<T>::infinity() doesn't run on the GPU
$ cat ./thrust/complex.h | grep numeric_limits  # nothing
$
```
If we were to rely on an external package manager instead of our own bundle to provide the complex headers, we wouldn't be able to fix it with your prescription.

**Comment by yxsamliu** [comment]
The issue can be demonstrated by https://godbolt.org/z/4rrsnfhe7

Both clang and gcc fails to compile it.

abs<thrust::complex<double>> is defined in the global name space as abs in https://github.com/llvm/llvm-project/blob/main/clang/lib/Headers/__clang_hip_cmath.h#L315. When clang compiles abs<thrust::complex<double>>, it needs evaluate numeric_limits<thrust::complex<double>> since it needs to determine whether abs<thrust::complex<double>> is a specialization of the abs template in __clang_hip_cmath.h, which causes an implicit instantiation of numeric_limits<thrust::complex<double>>.

Both __clang_cuda_cmath.h and __clang_hip_cmath.h use the similar approach to define math function templates, e.g.

https://github.com/llvm/llvm-project/blob/7e0cc45ced230b4ef3a9d8eaedfbe92e75f21916/clang/lib/Headers/__clang_cuda_cmath.h#L241

https://github.com/llvm/llvm-project/blame/7e0cc45ced230b4ef3a9d8eaedfbe92e75f21916/clang/lib/Headers/__clang_hip_cmath.h#L316

This puts restrictions on specialization of numeric_limits of user defined classes, i.e., they have to be defined before they are used.

what's interesting about this specific issue is that this regression seems to be caused by https://github.com/llvm/llvm-project/commit/aa2b593f1495a972a4a592952760ec9d5f7c01f1, where __HIP_OVERLOAD1(double, abs) is introduced, which was not there originally, also __clang_cuda_cmath.h does not have it. Therefore this regression only happens to HIP.

@JonChesterfield Do we really need __HIP_OVERLOAD1(double, abs) ?

**Comment by leofang** [comment]
Many thanks for pinpointing the issue, @yxsamliu! Would be great if this regression could be fixed in 4.1.x, or 4.2.0 if timeline does not permit. 🙂

**Comment by yxsamliu** [comment]
it is fixed in trunk https://reviews.llvm.org/D99738. not sure if we have enough time to get it in 4.2 but we will try

**Comment by leofang** [comment]
Fantastic, @yxsamliu!

**Comment by leofang** [comment]
This seems to be fixed indeed in ROCm 4.2.0 (cupy/cupy#5209). Thanks again @yxsamliu and the HIP team!
